app/vminsert/datadog/request_handler.go

```diff
@@ -17,70 +16,46 @@ var (
 	rowsPerInsert = metrics.NewHistogram(`vm_rows_per_insert{type="datadog"}`)
 )
 
-// InsertHandlerForHTTP processes remote write for DataDog POST /api/v1/series request.
+// InsertHandlerForHTTP processes remote write for DataDog POST /api/v1/series, /api/v2/series, /api/v1/sketches, /api/beta/sketches request.
 //
 // See https://docs.datadoghq.com/api/latest/metrics/#submit-metrics
 func InsertHandlerForHTTP(req *http.Request) error {
 	extraLabels, err := parserCommon.GetExtraLabels(req)
 	if err != nil {
 		return err
 	}
-	ce := req.Header.Get("Content-Encoding")
-	return stream.Parse(req.Body, ce, func(series []parser.Series) error {
-		return insertRows(series, extraLabels)
-	})
+	return stream.Parse(
+		req, func(series prompbmarshal.TimeSeries) error {
+			series.Labels = append(series.Labels, extraLabels...)
+			return insertRows(series)
+		},
+	)
 }
 
-func insertRows(series []parser.Series, extraLabels []prompbmarshal.Label) error {
+func insertRows(series prompbmarshal.TimeSeries) error {
 	ctx := common.GetInsertCtx()
 	defer common.PutInsertCtx(ctx)
 
-	rowsLen := 0
-	for i := range series {
-		rowsLen += len(series[i].Points)
-	}
-	ctx.Reset(rowsLen)
-	rowsTotal := 0
 	hasRelabeling := relabel.HasRelabeling()
-	for i := range series {
-		ss := &series[i]
-		rowsTotal += len(ss.Points)
-		ctx.Labels = ctx.Labels[:0]
-		ctx.AddLabel("", ss.Metric)
-		if ss.Host != "" {
-			ctx.AddLabel("host", ss.Host)
-		}
-		if ss.Device != "" {
-			ctx.AddLabel("device", ss.Device)
-		}
-		for _, tag := range ss.Tags {
-			name, value := parser.SplitTag(tag)
-			if name == "host" {
-				name = "exported_host"
-			}
-			ctx.AddLabel(name, value)
-		}
-		for j := range extraLabels {
-			label := &extraLabels[j]
-			ctx.AddLabel(label.Name, label.Value)
-		}
-		if hasRelabeling {
-			ctx.ApplyRelabeling()
-		}
-		if len(ctx.Labels) == 0 {
-			// Skip metric without labels.
-			continue
-		}
-		ctx.SortLabelsIfNeeded()
-		var metricNameRaw []byte
-		var err error
-		for _, pt := range ss.Points {
-			timestamp := pt.Timestamp()
-			value := pt.Value()
-			metricNameRaw, err = ctx.WriteDataPointExt(metricNameRaw, ctx.Labels, timestamp, value)
-			if err != nil {
-				return err
-			}
+	rowsTotal := len(series.Samples)
+
+	ctx.Reset(rowsTotal)
+	ctx.Labels = ctx.Labels[:0]
+	for l := range series.Labels {
+		ctx.AddLabel(series.Labels[l].Name, series.Labels[l].Value)
+	}
+	if hasRelabeling {
+		ctx.ApplyRelabeling()
+	}
+	if len(ctx.Labels) == 0 {
+		return nil
+	}
+	ctx.SortLabelsIfNeeded()
```

Comment:
From performance view it's better to create rawMetricName and re-use it for multiple samples:
```go

var metricNameRaw []byte

for _, sample := range series.Samples {
		metricNameRaw, err = ctx.WriteDataPointExt(metricNameRaw,ctx.Labels, sample.Timestamp, sample.Value)
}

Reply:
Get it, in this case it's better to replace slice allocation `[]byte{}` with `nil`.

Reply:
What about the following:
```
var metricNameRaw []byte

for _, sample := range series.Samples {
		metricNameRaw, err = ctx.WriteDataPointExt(metricNameRaw[:0],ctx.Labels, sample.Timestamp, sample.Value)
}
```
If we reset slice length we still can prevent extra allocations on each iteration.

Reply:
No, there'll be the same amount of allocation. function `ctx.marshalMetricNameRaw` uses own buffer for metricNameRaw and returns slice reference for it. So nil should be fine

Reply:
Ok. Please see this addressed here https://github.com/VictoriaMetrics/VictoriaMetrics/commit/98d0f81f21342296b20e13e2855dec9156c8b73c